driver-core/src/main/com/mongodb/internal/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -227,23 +226,29 @@ public void onResult(final List<RawBsonDocument> rawDocuments, final Throwable t
         void apply(AsyncAggregateResponseBatchCursor<RawBsonDocument> cursor, SingleResultCallback<List<RawBsonDocument>> callback);
     }
 
-    private void resumeableOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback,
+    private void resumeableOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> convertResultsCallback,
                                      final boolean tryNext) {
         if (isClosed()) {
-            callback.onResult(null, new MongoException(format("%s called after the cursor was closed.",
+            convertResultsCallback.onResult(null, new MongoException(format("%s called after the cursor was closed.",
                     tryNext ? "tryNext()" : "next()")));
             return;
         }
-        asyncBlock.apply(getWrapped(), new SingleResultCallback<List<RawBsonDocument>>() {
+        AsyncAggregateResponseBatchCursor<RawBsonDocument> wrappedCursor = getWrapped();
+        asyncBlock.apply(wrappedCursor, new SingleResultCallback<List<RawBsonDocument>>() {
             @Override
             public void onResult(final List<RawBsonDocument> result, final Throwable t) {
+                boolean retry = false;
                 if (t == null) {
-                    callback.onResult(result, null);
+                    convertResultsCallback.onResult(result, null);
                 } else if (isRetryableError(t, maxWireVersion)) {
-                    nullifyAndCloseWrapped();
-                    retryOperation(asyncBlock, callback, tryNext);
+                    retry = true;
                 } else {
-                    callback.onResult(null, t);
+                    convertResultsCallback.onResult(null, t);
+                }
+                cachePostBatchResumeToken(wrappedCursor);
```

Comment:
In the case where this method does call convertResultsCallback with a result, could it be consequential whether the post batch resume token is initialized prior to invoking the callback?  My concern is that the callback might be relying on it being set.  It's certainly strange that the order of operations is now:

1. set resumeToken to the last document's _id
2. invoke the callback
3. set resumeToken to the post batch resume token

Reply:
- Thanks! You are correct, the `set resumeToken to the post batch resume token` action must happen before we complete the callback just like `set resumeToken to the last document's _id` happens before completing the callback.
- On the other hand, `set resumeToken to the last document's _id` must happen before `set resumeToken to the post batch resume token` (as I described [here|https://github.com/mongodb/mongo-java-driver/pull/713#discussion_r632727156], though I am not sure is this is actually a "must").

These two constraints mean that with the current code I would have to call `cachePostBatchResumeToken` both

- in `convertResultsCallback` (right before calling `callback.onResult(results, null)` and `callback.onResult(null, null)`, i.e. in two places);
- and in `resumeableOperation` for the retry branch.

I will look if the code can be refactored to avoid such a mess.

Reply:
Done. There is no more callback that converts data, and even with the fix (it adds lines) the number of lines in `AsyncChangeStreamBatchCursor` went from 285 to 267, which I consider a small win :)

Reply:
Looks good!